File: app/src/main/java/com/letter/days/behavior/FabScrollBehavior.java
package com.letter.days.behavior;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.core.view.ViewCompat;
import androidx.appcompat.widget.Toolbar;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;

public class FabScrollBehavior extends CoordinatorLayout.Behavior {

    private boolean visible = true;

    public FabScrollBehavior (Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
    }

    @Override
    public boolean onStartNestedScroll(@NonNull CoordinatorLayout coordinatorLayout, @NonNull View child, @NonNull View directTargetChild, @NonNull View target, int axes, int type) {
        return axes == ViewCompat.SCROLL_AXIS_VERTICAL;
    }

    @Override
    public void onNestedScroll(@NonNull CoordinatorLayout coordinatorLayout, @NonNull View child, @NonNull View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed, int type) {
        super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, type);
        if (dyConsumed > 0 && visible) {
            visible = false;
            onHide(child);
        }
        else if (dyConsumed < 0) {
            visible = true;
            onShow(child);
        }
    }

    private void onHide(View view) {
        if (view instanceof Toolbar){
            ViewCompat.animate(view).translationY(-(view.getHeight() * 2)).setInterpolator(new AccelerateInterpolator(3));
        }else if (view instanceof FloatingActionButton){
            ViewCompat.animate(view).translationY(view.getHeight() * 2).setInterpolator(new AccelerateInterpolator(3));
        }
    }

    private void onShow(View view) {
        ViewCompat.animate(view).translationY(0).setInterpolator(new DecelerateInterpolator(3));

    }
}

File: app/src/main/java/com/letter/days/behavior/FabScrollBehavior.kt
package com.letter.days.behavior

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateInterpolator
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.ViewCompat


/**
 * Float Action Button 滑动行为
 * @param V: View 视图
 * @property visible Boolean 按钮可见
 * @constructor 构造器
 *
 * @author Letter(nevermindzzt@gmail.com)
 * @since 1.0.0
 */
class FabScrollBehavior<V: View>(context: Context, attrs: AttributeSet)
    : CoordinatorLayout.Behavior<V>(context, attrs) {

    private var visible = true

    override fun onStartNestedScroll(
        coordinatorLayout: CoordinatorLayout,
        child: V,
        directTargetChild: View,
        target: View,
        axes: Int,
        type: Int
    ) = (axes == ViewCompat.SCROLL_AXIS_VERTICAL)

    override fun onNestedScroll(
        coordinatorLayout: CoordinatorLayout,
        child: V,
        target: View,
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int,
        type: Int,
        consumed: IntArray
    ) {
        super.onNestedScroll(
            coordinatorLayout,
            child,
            target,
            dxConsumed,
            dyConsumed,
            dxUnconsumed,
            dyUnconsumed,
            type,
            consumed
        )
        if (dyConsumed > 0 && visible) {
            visible = false
            onHide(child)
        } else if (dyConsumed < 0) {
            visible = true
            onShow(child)
        }
    }

    private fun onHide(view: View) {
        ViewCompat.animate(view).translationY((view.height * 2).toFloat()).interpolator =
            AccelerateInterpolator(3F)
    }

    private fun onShow(view: View) {
        ViewCompat.animate(view).translationY(0F).interpolator =
            AccelerateInterpolator(3F)
    }
}